### PR TITLE
feat(appconfig): Automatically store "sensitive" appconfigs encrypted in the database

### PR DESCRIPTION
**Pull request by nickvergessen**
- [x] Based on #42903 

## Summary

Quite some app configs contain sensitive information that we could prevent to leak, when a backup of the DB is leaked, to reduce the risk of follow up leaks/hacks.
The state of the art is to use `ICrypto::encrypt()` and `ICrypto::decrypt()` for that. While looking into it for some of the talk configs which are arrays, I thought it's a shame that we would loose all the typing as we have to json_encode the config making it strings, in order to be able to encrypt them.

So why not just encrypt all sensitive configs all the time automatically?

## TODO

- [x] Adjust normal getter and setter
- [x] Check `getAllValues` and `getDetails`

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: lib/private/AppConfig.php [hunk 1]
```diff
@@ -462,11 +464,24 @@ private function getTypedValue(
 			throw new AppConfigTypeConflictException('conflict with value type from database');
 		}
 
-		if ($lazy) {
-			return $this->lazyCache[$app][$key] ?? $default;
+		if ($lazy && isset($this->lazyCache[$app][$key])) {
+			$value = $this->lazyCache[$app][$key];
```
Comment (ArtificialOwl) [inline lib/private/AppConfig.php:468]: what if you put a ref as `$value = &$this->lazyCache[$app][$key];` ? would it update the value in the global array with decrypted one ?

**Review by ArtificialOwl — APPROVED** [review]

File: lib/private/AppConfig.php [hunk 2]
```diff
@@ -462,11 +464,24 @@ private function getTypedValue(
...
+			return $default;
 		}
 
-		return $this->fastCache[$app][$key] ?? $default;
+		$sensitive = $this->isTyped(self::VALUE_SENSITIVE, $knownType);
+		if ($sensitive) {
+			// Only decrypt values that are stored encrypted
+			$sections = substr_count($value, '|');
+			if ($sections === 2 || $sections === 3) {
+				$value = $this->crypto->decrypt($value);
```
Comment (ArtificialOwl) [inline lib/private/AppConfig.php:480]: what would happen if unencrypted sensitive string fullfil this condition ?
should we add a prefix to the stored value to help identify encrypted value ?
do we manage/ignore exception on decrypt ?

File: lib/private/AppConfig.php (hunk 1 above)
Comment (nickvergessen) [reply lib/private/AppConfig.php:468]: I dislike such side-effects, but anyway I iirc references are not working for scalar types in php

File: lib/private/AppConfig.php [hunk 3]
```diff
@@ -728,12 +743,17 @@ private function setTypedValue(
...
-		if ($this->hasKey($app, $key, $lazy)
+		if (!$sensitive
+			&& $this->hasKey($app, $key, $lazy)
 			&& $value === $this->getTypedValue($app, $key, $value, $lazy, $type)
-			&& (!$sensitive || $this->isSensitive($app, $key, $lazy))) {
+			&& !$this->isSensitive($app, $key, $lazy)) {
 			return false;
 		}
 
+		if ($sensitive) {
```
Comment (ArtificialOwl) [inline lib/private/AppConfig.php:753]: firefox crashed and seems suggestion was not attached to my previous review ...

`$sensitive` can be false, but once set as sensitive in database it won't be reverted unless using `updateSensitive()`

```suggestion
		if ($sensitive || $this->isSensitive($app, $key, $lazy)) {
```

File: lib/private/AppConfig.php [hunk 4]
```diff
@@ -728,12 +743,17 @@ private function setTypedValue(
 		 * no update if key is already known with set lazy status, or value is
 		 * different, or sensitivity switched from false to true.
 		 */
-		if ($this->hasKey($app, $key, $lazy)
+		if (!$sensitive
+			&& $this->hasKey($app, $key, $lazy)
 			&& $value === $this->getTypedValue($app, $key, $value, $lazy, $type)
-			&& (!$sensitive || $this->isSensitive($app, $key, $lazy))) {
+			&& !$this->isSensitive($app, $key, $lazy)) {
```
[the remark this answers is not in the document]
Comment (nickvergessen) [reply lib/private/AppConfig.php:749]: > meaning, while new and old value are the same, we keep update the database if $sensitive is true but value in database is not yet set as sensitive !

That's exactly what I tried to do

File: lib/private/AppConfig.php (hunk 3 above)
Comment (nickvergessen) [reply lib/private/AppConfig.php:752]: Shouldn't `$sensitive = $this->isTyped(self::VALUE_SENSITIVE, $type);` above then have this?
Also we would then need to manipulate the type and make sure we store it, even when it was not provided.

All those methods are quite complex and all invoking each others and it's really not straight forward atm for me to see what is right.

Comment (nickvergessen) [reply lib/private/AppConfig.php]: Got it now, and added it in a second commit

**Review by ArtificialOwl — APPROVED** [review]

**Review by AndyScherzinger — APPROVED** [review]

**Pull request merged**